File: ios/RNSPUserData/RNSPPreferencesConsent.swift
import ConsentViewController

struct RNSPPreferencesConsent: Encodable {
  struct Status: Encodable {
    let categoryId: Int
    let channels: [Channel]
    let changed: Bool?
    let dateConsented: SPDate?
    let subType: SubType?
  }

  struct Channel: Encodable {
    let id: Int
    let status: Bool
  }

  enum SubType: String, Encodable {
    case AIPolicy, TermsAndConditions, PrivacyPolicy, LegalPolicy, TermsOfSale, Unknown

    init(from subType: SPPreferencesConsent.SubType) {
      switch subType {
      case .AIPolicy: self = .AIPolicy
      case .TermsAndConditions: self = .TermsAndConditions
      case .PrivacyPolicy: self = .PrivacyPolicy
      case .LegalPolicy: self = .LegalPolicy
      case .TermsOfSale: self = .TermsOfSale
      default: self = .Unknown
      }
    }
  }

  let dateCreated: SPDate
  let uuid: String?
  let status: [Status]
  let rejectedStatus: [Status]
}

extension RNSPPreferencesConsent.Status {
  init(from status: SPPreferencesConsent.Status) {
    categoryId = status.categoryId
    channels = status.channels.map { RNSPPreferencesConsent.Channel(id: $0.id, status: $0.status) }
    changed = status.changed
    dateConsented = status.dateConsented
    subType = status.subType.flatMap { RNSPPreferencesConsent.SubType(from: $0) }
  }
}

extension RNSPPreferencesConsent.Channel {
  init(from channel: SPPreferencesConsent.Channel) {
    id = channel.id
    status = channel.status
  }
}

extension RNSPPreferencesConsent {
  init?(from preferences: SPPreferencesConsent?) {
    guard let preferences = preferences else { return nil }

    self.init(
      dateCreated: preferences.dateCreated,
      uuid: preferences.uuid,
      status: preferences.status.map { Status(from: $0) },
      rejectedStatus: preferences.rejectedStatus.map { Status(from: $0) }
    )
  }
}

File: android/src/main/java/com/sourcepoint/reactnativecmp/consents/RNSPPreferencesConsent.kt
package com.sourcepoint.reactnativecmp.consents

import com.facebook.react.bridge.Arguments.createArray
import com.facebook.react.bridge.Arguments.createMap
import com.facebook.react.bridge.ReadableMap
import com.sourcepoint.cmplibrary.model.exposed.PreferencesConsent

data class RNSPPreferencesConsent(
  val dateCreated: String?,
  val uuid: String?,
  val status: List<Status>,
  val rejectedStatus: List<Status>
): RNMappable {
  data class Status(
    val categoryId: Int,
    val channels: List<Channel>,
    val changed: Boolean?,
    val dateConsented: String?,
    val subType: String?
  ): RNMappable {
    data class Channel(val id: Int, val status: Boolean): RNMappable {
      override fun toRN(): ReadableMap = createMap().apply {
        putInt("id", id)
        putBoolean("status", status)
      }
    }

    override fun toRN(): ReadableMap = createMap().apply {
      putInt("categoryId", categoryId)
      putArray("channels", createArray().apply { channels.forEach { pushMap(it.toRN()) } })
      changed?.let { putBoolean("changed", it) }
      putString("dateConsented", dateConsented)
      putString("subType", subType)
    }
  }

  constructor(preferences: PreferencesConsent) : this(
    dateCreated = preferences.dateCreated?.toString(),
    uuid = preferences.uuid,
    status = preferences.status?.map { status ->
      Status(
        categoryId = status.categoryId,
        channels = status.channels?.map { channel ->
          Status.Channel(id = channel.id, status = channel.status)
        } ?: emptyList(),
        changed = status.changed,
        dateConsented = status.dateConsented?.toString(),
        subType = status.subType?.toString()
      )
    } ?: emptyList(),
    rejectedStatus = preferences.rejectedStatus?.map { status ->
      Status(
        categoryId = status.categoryId,
        channels = status.channels?.map { channel ->
          Status.Channel(id = channel.id, status = channel.status)
        } ?: emptyList(),
        changed = status.changed,
        dateConsented = status.dateConsented?.toString(),
        subType = status.subType?.toString()
      )
    } ?: emptyList()
  )

  override fun toRN(): ReadableMap = createMap().apply {
    putString("dateCreated", dateCreated)
    putString("uuid", uuid)
    putArray("status", createArray().apply { status.forEach { pushMap(it.toRN()) } })
    putArray("rejectedStatus", createArray().apply { rejectedStatus.forEach { pushMap(it.toRN()) } })
  }
}
